### ISSUE TITLE
Incompatibilities between Indic and USE cluster validation

### ISSUE THREAD

**Issue by dscorbett** [harfbuzz/harfbuzz#1746]
A string in a pre-USE Indic script should be renderable with either an Indic2 or Indic3 font. However, the Indic and USE shapers use different Ragel machines, so they disagree about which strings are valid. Some features of the Indic machine should be ported to the USE machine for compatibility. Which and how?

For example, the Indic machine allows a vowel sign to precede a nukta. The USE machine does not. It makes sense that a spacing (i.e. pre- or post-base) vowel sign could take a nukta, and in fact [it is attested in the Tamil script](https://www.unicode.org/L2/L2015/15256-tamil-nukta.pdf), so I recommend changing USE’s VPre\* VAbv\* VBlw\* VPst\* to (VPre CMAbv\* CMBlw\*)\* VAbv\* VBlw\* (VPst CMAbv\* CMBlw\*)\*. Still, that is not completely compatible with the Indic machine.

Another example is `forced_rakar`, which was introduced for Sinhala in commit 422ecd2d3c198a36d07d409341cb82ea57c7ad6b. Sinhala doesn’t even use the Indic shaper in Windows 10, so maybe this hack is unnecessary. I recommend not porting it to USE.

When I added support for parsing Indic3 tags, I included 'mym3' because it was easier to implement that way. However, 'mym2' uses its own, very different machine, and I shouldn’t have assumed it would have an Indic3 variant. Does Apple support 'mym3'? Supporting Myanmar in USE would require many script-specific overrides and categories; it would be even worse than Tai Tham. So I’d rather drop 'mym3' if possible.

There are plenty of other examples. Without yet focusing too much on any particular incompatibility, what is the best approach here? How much compatibility with the Indic machine is necessary? Should the changes be restricted to the Indic2 scripts or available to all USE scripts?

**Comment by behdad** [comment]
These are very good points.  I'll bring them up with Andrew Glass et al next week.  Thanks.

**Comment by behdad** [comment]
@dscorbett I sent you an email.  Can you please reply.  Or if you didn't receive it, send me your email address.  Thanks.

**Comment by behdad** [comment]
Is this actionable? @dscorbett

**Comment by dscorbett** [comment]
Yes. The first step is to decide the general approach: targeted exceptions for compatibility vs. broad changes for consistency. The next step is to find and document all the discrepancies between the two shapers. Ultimately it would be nice if the Indic shaper could be deleted and everything could use USE without changing the rendering of any existing valid Indic text.
